onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorShape.cpp

```diff
@@ -0,0 +1,100 @@
+    :   DmlOperator(kernelCreationContext),
+        ShapeHelper(kernelCreationContext, kernelCreationContext.GetTensorShapeDescription())
+    {
+        ML_CHECK_VALID_ARGUMENT(kernelCreationContext.GetInputCount() == 1, "Shape expects 1 input tensor.");
+        ML_CHECK_VALID_ARGUMENT(kernelCreationContext.GetOutputCount() == 1, "Shape expects 1 output tensor.");
+
+        m_startIndex = kernelCreationContext.GetOptionalAttribute<int32_t>(AttrName::Start, 0);
+
+        if (m_startIndex != 0)
+        {
+            // "start" is provided and is non-default (default is 0)
+            m_needsSlicing = true;
+        }
+
+        if (kernelCreationContext.HasAttribute(AttrName::End, MLOperatorAttributeType::Int))
+        {
+            m_endIndex = kernelCreationContext.GetAttribute<int64_t>(AttrName::End);
+            m_needsSlicing = true;
+        }
+
+        DmlOperator::Initialize(kernelCreationContext);
+    }
+
+    // Takes a tensor as input and outputs a 1D int64 tensor containing the shape of the input tensor.
+    void Compute(const MLOperatorKernelContext& kernelContext)
+    {
+        std::vector<IMLOperatorTensor*> inputTensors = GetInputTensors(kernelContext);
+        std::vector<IMLOperatorTensor*> outputTensors = GetOutputTensors(kernelContext);
+        assert(inputTensors.size() == 1);
+        assert(outputTensors.size() == 1);
+
+        const IMLOperatorTensor* inputTensor = inputTensors[0];
+        IMLOperatorTensor* outputTensor = outputTensors[0];
+        assert(outputTensor->GetDimensionCount() == 1);
+
+        uint32_t inputDimCount = inputTensor->GetDimensionCount();
+        uint32_t outputSize;
+        THROW_IF_FAILED(outputTensor->GetShape(1, &outputSize));
+        assert(outputSize == inputDimCount);
+
+        // Get output shape as a vector
+        std::vector<uint32_t> outputShape;
+        if (!m_needsSlicing)
+        {
+            outputShape.resize(inputDimCount);
+            THROW_IF_FAILED(inputTensor->GetShape(inputDimCount, outputShape.data()));
+        }
+        else
+        {
+            int64_t trueStart = m_startIndex;
+            int64_t trueEnd = m_endIndex;
+
+            // Deal with negative(s) and clamp
+            trueStart = trueStart < 0 ? trueStart + inputDimCount : trueStart;
+            trueStart = trueStart < 0 ? 0 : ((trueStart > inputDimCount) ? inputDimCount : trueStart);
+
+            trueEnd = trueEnd < 0 ? trueEnd + inputDimCount : trueEnd;
+            trueEnd = trueEnd < 0 ? 0 : ((trueEnd > inputDimCount) ? inputDimCount : trueEnd);
+
+            int64_t sliceLength = trueEnd - trueStart;
+            outputShape.resize(sliceLength);
+
+            if (sliceLength > 0)
+            {
+                THROW_IF_FAILED(inputTensor->GetShape(static_cast<uint32_t>(sliceLength), outputShape.data() + static_cast<uint32_t>(trueStart)));
+            }
+        }
+
+        int64_t* outputData = reinterpret_cast<int64_t*>(outputTensor->GetData());
+
```

Comment:
Repeating comment from Size:

A sanity check here would be nice before writing directly to memory without checking the EP or byte size, like ReadScalarTensorData checking IsCpuData and the byte size. (it should be correct by the time it reaches here 🤞, but distant sections of code change over time).